main/src/cgeo/geocaching/ui/dialog/CoordinatesCalculateDialog.java

```diff
@@ -0,0 +1,1096 @@
+
+import android.app.Dialog;
+import android.os.Bundle;
+import android.support.annotation.IdRes;
+import android.support.v4.BuildConfig;
+import android.support.v4.app.DialogFragment;
+import android.support.v7.widget.GridLayout;
+import android.text.Editable;
+import android.text.TextWatcher;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.view.Window;
+import android.widget.AdapterView;
+import android.widget.ArrayAdapter;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.HorizontalScrollView;
+import android.widget.ImageButton;
+import android.widget.Spinner;
+import android.widget.TextView;
+
+import org.apache.commons.lang3.SerializationUtils;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import butterknife.ButterKnife;
+import cgeo.geocaching.R;
+import cgeo.geocaching.activity.AbstractActivity;
+import cgeo.geocaching.location.Geopoint;
+import cgeo.geocaching.location.GeopointFormatter;
+import cgeo.geocaching.models.CalcState;
+import cgeo.geocaching.sensors.Sensors;
+import cgeo.geocaching.settings.Settings;
+import cgeo.geocaching.ui.CalculateButton;
+import cgeo.geocaching.ui.EditButton;
+import cgeo.geocaching.ui.CalculatorVariable;
+
+import static cgeo.geocaching.R.id.PlainFormat;
+import static cgeo.geocaching.R.id.coordTable;
+import static cgeo.geocaching.models.CalcState.ERROR_CHAR;
+
+public class CoordinatesCalculateDialog extends DialogFragment {
+
+    private static final String PLACE_HOLDER = "~"; // Character used to represent a "blanked-out' CoordinateButton.
+
+    static final String SYMBOL_DEG = "°";
+    static final String SYMBOL_MIN = "'";
+    static final String SYMBOL_SEC = "\"";
+    static final String SYMBOL_POINT = ".";
+
+    private Geopoint gp;
+    private CalcState calcState;
+
+    // Through out this implementation:
+    //
+    // 'Equations' are used to represent 'Variables' that appear in the description of the cache coordinated themselves.
+    //             As in "N 42° 127.ABC".  In this example 'A', 'B' and 'C' are all 'equations'.
+    //             All 'equations' must have a CAPITAL-LETTER name.
+    //
+    // 'FreeVariables' are used to represent 'Variables' that appear in the 'expression' of an equation
+    //                 As in "X = a^2 + b^2".  In this example 'a' and 'b' are both 'freeVariables'.
+    //                 All 'freeVariables' must have a lower-case name.
+
+    List<CalculatorVariable> equations;       // List of equations to be displayed in the calculator.
+    List<CalculatorVariable> freeVariables;   // List of freeVariables to be displayed in the calculator.
+
+    Spinner spinner;
+
+    private String inputLatHem;
+    private Button bLatHem;  // Latitude hemisphere (North or South)
+    private CalculateButton         bLatDeg_100,   bLatDeg_010,   bLatDeg_001,

```

Comment:
As I suggested in the other pull request: Can you think of a better way to store these variables? They are already formatted as an array, so you could either try `CalculateButton[][]` or a custom class.